# learn-genomics-in-unix

This repository is an introductory tutorial on how to use command-line genomics/bioinformatics tools in Linux.

It is divided into a series of tasks

  * Task 1 - Learning the linux command line
  * Task 1b - Command-line BLAST
  * Task 2 - Genome Assembly
  * Task 3 - Genome Annotation
  * Task 4 - Comparative genomics: synteny comparison between two genomes
  * Task 5 - Comparative genomics: gene set comparison between two genomes
  * Task 6 - Metagenomics: taxonomic and functional profiling
  * Task 7 - Resequencing: variant calling from NGS data
  
# Requirements & Software Installation

The main requirement is that you have access to a Linux-based OS such as Ubuntu.
Most of the programs we will use are text-based and can be run directly in the shell; however, some graphical programs (e.g., Tablet, Artemis, fastqc) will be used as well. You can install these on your own local machine.

Detailed software requirements will be listed at the beginning of each Task. Alternatively, if you wish to install all the programs for the course beforehand, please install the software listed below:

* [tablet](https://ics.hutton.ac.uk/tablet/) *
* [artemis and act](http://sanger-pathogens.github.io/Artemis/Artemis/) *
* [R](https://www.r-project.org/)

We have also documented installation instructions for building a bioinformatics system (Ubuntu 18.04) [here](https://github.com/doxeylab/learn-genomics-in-unix/blob/master/VMbuild.README).


 
# Development & Contact

This tutorial was developed by Dr. Andrew Doxey and other members of the Doxey Lab (http://doxey.uwaterloo.ca/).
If you have any questions, please contact acdoxey at uwaterloo dot ca.

Enjoy
